packages/runtime/runtime-utils/src/summarizerNodeUtils.ts

```diff
@@ -0,0 +1,253 @@
+/*!
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { ITelemetryLogger } from "@fluidframework/common-definitions";
+import { assert } from "@fluidframework/common-utils";
+import {
+    ISnapshotTree,
+    IDocumentAttributes,
+    ISequencedDocumentMessage,
+    SummaryType,
+} from "@fluidframework/protocol-definitions";
+import { ISummaryTreeWithStats } from "@fluidframework/runtime-definitions";
+import { SummaryTreeBuilder } from "./summaryUtils";
+
+const baseSummaryTreeKey = "_baseSummary";
+const outstandingOpsBlobKey = "_outstandingOps";
+const maxDecodeDepth = 100;
+
+/** Reads a blob from storage and parses it from JSON. */
+export type ReadAndParseBlob = <T>(id: string) => Promise<T>;
+
+/**
+ * Fetches the sequence number of the snapshot tree by examining the protocol.
+ * @param tree - snapshot tree to examine
+ * @param readAndParseBlob - function to read blob contents from storage
+ * and parse the result from JSON.
+ */
+export async function seqFromTree(
+    tree: ISnapshotTree,
+    readAndParseBlob: ReadAndParseBlob,
```

Comment:
Can you put the generic type parameter here? As-is it's basically no better than an `as` cast, or so it seems to me.

Reply:
Not sure what exactly you mean by this.

I promote the type parameter like this:
```typescript
export type ReadAndParseBlob<T> = (id: string) => Promise<T>;
...
    readAndParseBlob: ReadAndParseBlob<IDocumentAttributes>,
...
    const attrib = await readAndParseBlob<(attributesHash);
```

But I'm not really sure what benefit that brings me.
The contract now is: the caller gives me a function which reads and parses some blobs to whatever type I think each one is.
The contract after this change is: the caller gives me a function which reads and parses some blobs to IDocumentAttributes.

While the 2nd is narrower, I'm not sure it's actually more accurate.

Reply:
Ok I'll leave it up to your judgment.  But the caller giving a function that doesn't know what type it's going to return is surprising to me.